indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/PendingTaskBasedResourceManagementStrategy.java

```diff
@@ -0,0 +1,360 @@
+import io.druid.indexing.overlord.setup.WorkerBehaviorConfig;
+import io.druid.indexing.overlord.setup.WorkerSelectStrategy;
+import io.druid.indexing.worker.Worker;
+import org.joda.time.DateTime;
+import org.joda.time.Duration;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ */
+public class PendingTaskBasedResourceManagementStrategy implements ResourceManagementStrategy
+{
+  private static final EmittingLogger log = new EmittingLogger(PendingTaskBasedResourceManagementStrategy.class);
+
+  private final PendingTaskBasedResourceManagementConfig config;
+  private final Supplier<WorkerBehaviorConfig> workerConfigRef;
+  private final ScalingStats scalingStats;
+
+  private final Object lock = new Object();
+  private final Set<String> currentlyProvisioning = Sets.newHashSet();
+  private final Set<String> currentlyTerminating = Sets.newHashSet();
+
+  private DateTime lastProvisionTime = new DateTime();
+  private DateTime lastTerminateTime = new DateTime();
+
+  @Inject
+  public PendingTaskBasedResourceManagementStrategy(
+      PendingTaskBasedResourceManagementConfig config,
+      Supplier<WorkerBehaviorConfig> workerConfigRef
+  )
+  {
+    this.config = config;
+    this.workerConfigRef = workerConfigRef;
+    this.scalingStats = new ScalingStats(config.getNumEventsToTrack());
+  }
+
+  @Override
+  public boolean doProvision(RemoteTaskRunner runner)
+  {
+    Collection<Task> pendingTasks = runner.getPendingTaskPayloads();
+    Collection<ZkWorker> zkWorkers = runner.getWorkers();
+    synchronized (lock) {
+      boolean didProvision = false;
+      final WorkerBehaviorConfig workerConfig = workerConfigRef.get();
+      if (workerConfig == null || workerConfig.getAutoScaler() == null) {
+        log.warn("No workerConfig available, cannot provision new workers.");
+        return false;
+      }
+
+      final List<String> workerNodeIds = workerConfig.getAutoScaler().ipToIdLookup(
+          Lists.newArrayList(
+              Iterables.<ZkWorker, String>transform(
+                  zkWorkers,
+                  new Function<ZkWorker, String>()
+                  {
+                    @Override
+                    public String apply(ZkWorker input)
+                    {
+                      return input.getWorker().getIp();
+                    }
+                  }
+              )
+          )
+      );
+      currentlyProvisioning.removeAll(workerNodeIds);
+      if (currentlyProvisioning.isEmpty()) {
+        int want = getScaleUpNodeCount(
+            runner.getConfig(),
+            workerConfig,
+            pendingTasks,
+            zkWorkers
+        );
+        while (want > 0) {
+          final AutoScalingData provisioned = workerConfig.getAutoScaler().provision();
+          final List<String> newNodes;
+          if (provisioned == null || (newNodes = provisioned.getNodeIds()).isEmpty()) {
```

Comment:
Can this logic be moved to newNodes for something like `newNodes = provisioned == null ? ImmutableList.of() : provisioned.getNodeIds()` with this check being on `newNodes.isEmpty()`? I think that will make it a bit easier to follow the logic.